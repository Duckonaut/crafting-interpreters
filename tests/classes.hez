class Character {
    init(name) {
        self.name = name;
    }

    sayHi() {
        print("Hi, I'm ");
        println(self.name);
    }

    haveAChild() {
        return Character("Child of " + self.name);
    }

    getFlag() {
        class Flag {
            init(name) {
                self.name = name;
            }

            wave() {
                print("The flag of ");
                println(self.name);
            }
        }

        return Flag(self.name);
    }

    getFlagWaveCallback() {
        return self.getFlag().wave;
    }
}

class Mario < Character {
    init() {
        super.init("Mario");
    }

    jump() {
        println("Jump!");
    }
}
println("defined");

var mario = Mario();
var koopa = Character("Koopa");

mario.jump();
println(mario.haveAChild().name);

var childWaveLater = mario.haveAChild().getFlagWaveCallback();
var marioFlag = mario.getFlag();

marioFlag.wave();
childWaveLater();
