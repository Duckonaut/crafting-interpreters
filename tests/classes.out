defined
Jump!
Child of Mario
The flag of Mario
The flag of Child of Mario

